src/google/protobuf/compiler/objectivec/objectivec_file.cc

```diff
@@ -84,6 +84,52 @@ bool MessageContainsExtensions(const Descriptor* message) {
   return false;
 }
 
+// Checks if the Message contains a field that requires the custom GPBArray
+// types. Checks at this message and sub messages.
+bool MessageContainsGPBArrayField(const Descriptor* message) {
+  for (int i = 0; i < message->field_count(); i++) {
+    const FieldDescriptor* field = message->field(i);
+    if (field->is_repeated() && !field->is_map() && IsPrimitiveType(field)) {
+      return true;
+    }
+  }
+
+  for (int i = 0; i < message->nested_type_count(); i++) {
+    if (MessageContainsGPBArrayField(message->nested_type(i))) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Checks if the Message contains a field that requires the custom
+// GPBDictionary types. Checks at this message and sub messages.
+bool MessageContainsGPBDictField(const Descriptor* message) {
```

Comment:
MessageContainsMapField? the GPBDict naming is for objc protos, and technically we're dealing with proto descriptors here

Reply:
Because some map<> fields just use NSMutableDictionary (String->Message, String->Data, String->String).  So this really is a test to see if they will need a GPBDict class.

Reply:
Same goes for the Array side, String/Data/Message all use NSMutableArray, so it was being specific to what gets the custom class.